### PR TITLE
rustdoc: Fix generics generation in search index

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
The generics were not added to the search index as they should, instead they were added as arguments. I used this opportunity to allow generics to have generics themselves (will come in very handy for my current rewrite of the search engine!).

r? @jyn514

### CODE REVIEW

**Review by jyn514 — COMMENTED** [review]
> I used this opportunity to allow generics to have generics themselves

What does this mean? What code does it allow you to represent? I'm not really clear what any of this code is doing ...

File: src/librustdoc/formats/item_type.rs [hunk 1]
```diff
@@ -147,32 +148,33 @@ impl From<DefKind> for ItemType {
...
-            ItemType::Constant => "constant",
-            ItemType::AssocConst => "associatedconstant",
-            ItemType::ForeignType => "foreigntype",
-            ItemType::Keyword => "keyword",
-            ItemType::OpaqueTy => "opaque",
-            ItemType::ProcAttribute => "attr",
-            ItemType::ProcDerive => "derive",
-            ItemType::TraitAlias => "traitalias",
+            Self::Module => "mod",
+            Self::ExternCrate => "externcrate",
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/formats/item_type.rs:152]: Arf, I didn't even pay attention... I'll revert this change.

File: src/librustdoc/html/render/cache.rs [hunk 2]
```diff
@@ -194,32 +194,24 @@ crate fn get_index_search_type<'tcx>(
...
 
     Some(IndexItemFunctionType { inputs, output })
 }
 
-fn get_index_type(clean_type: &clean::Type) -> RenderType {
+fn get_index_type(clean_type: &clean::Type, generics: Vec<TypeWithKind>) -> RenderType {
     RenderType {
         name: get_index_type_name(clean_type, true).map(|s| s.as_str().to_ascii_lowercase()),
-        generics: get_generics(clean_type),
+        generics: if generics.is_empty() { None } else { Some(generics) },
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:214]: Because if it's not an option, it'll generate `[]`, which is unneeded. Before to not generate anything then.

**Comment by GuillaumeGomez** [comment]
> > I used this opportunity to allow generics to have generics themselves
>
> What does this mean? What code does it allow you to represent? I'm not really clear what any of this code is doing ...

Let's take an example then!

```rust
struct Foo<A, B>(A, B);

fn foo<A: Display, B: Foo<u32, u64>, C: Foo<A, B>>(arg: A) {}
```

Like this, `arg` will be a generic `A` which itself contains the generics `B` and `C` which themselves have generics. Before that, it was all inlined as arguments of the functions, which is at best incorrect.

File: src/librustdoc/html/render/cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:214]: I'll add a comment to make it clear.

File: src/librustdoc/html/render/cache.rs [hunk 3]
```diff
@@ -278,27 +253,33 @@ crate fn get_real_types<'tcx>(
...
+            res.push(TypeWithKind::from((index_ty, ItemType::Generic)));
+        } else if let Some(kind) = ty.def_id().map(|did| tcx.def_kind(did).into()) {
+            res.push(TypeWithKind::from((index_ty, kind)));
         } else if ty.is_primitive() {
             // This is a primitive, let's store it as such.
-            res.insert((ty, ItemType::Primitive));
-            1
-        } else {
-            0
+            res.push(TypeWithKind::from((index_ty, ItemType::Primitive)));
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:275]: There is a `ty.is_primitive()` check just like before. Did I miss something?

[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/html/render/cache.rs:319]: Oh, I misread the code. Why did you remove the else branch? What did 0 represent?

File: src/librustdoc/html/render/cache.rs [hunk 4]
```diff
@@ -278,33 +253,84 @@ crate fn get_real_types<'tcx>(
...
+            //
+            // After removing the intermediate (unnecessary) full generic, it'll become:
+            //
+            // ```
+            // [{
+            //     name: "option",
+            //     generics: [{
+            //         name: "Display",
+            //         generics: []
+            //     ]
```
Comment (jyn514) [inline src/librustdoc/html/render/cache.rs:298]:
```suggestion
            //     }]
```

File: src/librustdoc/html/render/cache.rs [hunk 5]
```diff
@@ -278,33 +253,84 @@ crate fn get_real_types<'tcx>(
...
+            // }]
+            // ```
+            //
+            // To be noted that it can work if there is ONLY ONE generic, otherwise we still
```
Comment (jyn514) [inline src/librustdoc/html/render/cache.rs:302]: Why does it only work if there's a single generic argument? It seems weird to have an outer object that just has `name: ""`

File: src/librustdoc/json/conversions.rs [hunk 6]
```diff
@@ -677,6 +677,7 @@ impl FromWithTcx<ItemType> for ItemKind {
             TraitAlias => ItemKind::TraitAlias,
             ProcAttribute => ItemKind::ProcAttribute,
             ProcDerive => ItemKind::ProcDerive,
+            _ => unreachable!(),
```
Comment (jyn514) [inline src/librustdoc/json/conversions.rs:680]: Please use the name instead of a wildcard.
```suggestion
            Generic => unreachable!(),
```

File: src/librustdoc/html/render/cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/html/render/cache.rs:214]: I'm confused though - can't you just do this same check wherever it would generate an empty Vec? Why does it have to be specifically here?

**Review by jyn514 — COMMENTED** [review]
Is there a reference for how the search index works somewhere? Your examples helped but I have no idea how they'll actually affect search, I don't really feel comfortable signing off on this. Maybe @notriddle or @ollie27 are familiar with search?

File: src/test/rustdoc-js/generics.js [hunk 7]
```diff
@@ -3,7 +3,6 @@
 const QUERY = [
   '"R<P>"',
   '"P"',
-  'P',
```
Comment (notriddle) [inline src/test/rustdoc-js/generics.js:6]: It seems odd that you’re removing the non-quoted search, but not removing the quoted search.

File: src/librustdoc/html/render/cache.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:319]: Before it returned the number of items inserted to know if something needed to be added into the HashSet. Now we add all arguments (as long as they have a name).

File: src/librustdoc/html/render/cache.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:302]: It's what I tried to explain but I failed I guess haha.

So basically, arguments can only have a type and generics. What happens in case the argument is a full generic which is `Display + Debug + Write`? Then we can't say it's one type (from a search perspective) but a generic type (with empty name) having three generics.

File: src/librustdoc/json/conversions.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply src/librustdoc/json/conversions.rs:680]: :+1:

File: src/librustdoc/html/render/cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:214]: I'm not sure about the other places... In this case, it's contained inside a vec, so if the vec length is 1 or 2, it's not really a problem. For the other places, it might be different.

File: src/test/rustdoc-js/generics.js (hunk 7 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc-js/generics.js]: I updated the test to take the update into account.

**Comment by GuillaumeGomez** [comment]
ping @jyn514

**Comment by GuillaumeGomez** [comment]
I think @jyn514 and I need this PR to be merged to end the rewrite of the rustdoc search engine so I'll set @camelid as reviewer.

r? @camelid

**Comment by camelid** [comment]
> > > I used this opportunity to allow generics to have generics themselves
> > 
> > 
> > What does this mean? What code does it allow you to represent? I'm not really clear what any of this code is doing ...
> 
> Let's take an example then!
> 
> ```rust
> struct Foo<A, B>(A, B);
> 
> fn foo<A: Display, B: Foo<u32, u64>, C: Foo<A, B>>(arg: A) {}
> ```
> 
> Like this, `arg` will be a generic `A` which itself contains the generics `B` and `C` which themselves have generics. Before that, it was all inlined as arguments of the functions, which is at best incorrect.

I don't really understand this example. Why will `A` contain `B` and `C`? Is it some kind of linked-list/tree structure? And if so, why?

Also, `Foo` is a struct, so it can't be a trait bound. I assume you meant something like `trait Foo<A, B> {}`?

How did the search index look previously? Based on your description, it sounds like the type of this function used to be treated as something like `fn(Display, Foo<u32, u64>, Foo<A, B>, A)`? Is that correct? And what will it look like after this change?

Sorry for all the questions; I'm not very familiar with the search-index code ;)

**Comment by GuillaumeGomez** [comment]
Woups, I meant that `C` contains `A` and `B`. It should have been:

```rust
struct Foo<A, B>(A, B);

fn foo<A: Display, B: Foo<u32, u64>, C: Foo<A, B>>(arg: C) {}
```

And indeed, it looks like a tree:

```
arg: [
    {
        type: "Foo",
        generics: [
            {
                type: "Display",
                generics: []
            },
            {
                type: "Foo"
                generics: [
                    {
                        type: "u32",
                        generics: []
                    },
                    {
                        type: "u64",
                        generics: []
                    }
                ]
            }
        ]
    }
]
```

File: src/librustdoc/formats/item_type.rs [hunk 8]
```diff
@@ -48,6 +48,7 @@ crate enum ItemType {
     ProcAttribute = 23,
     ProcDerive = 24,
     TraitAlias = 25,
+    Generic = 26,
```
Comment (camelid) [inline src/librustdoc/formats/item_type.rs:51]: Why are generic parameters treated as items?

Comment (GuillaumeGomez) [reply src/librustdoc/formats/item_type.rs:51]: The pure generic ones, like `fn foo<T: Display + Debug>(a: T)`, in here, `a` cannot be both "Display" and "Debug", it needs to go through a level:

```
arg: [
  {
    name: "",
    generics: [
      {
        name: "Display"
        generics: []
      },
      {
        name: "Debug"
        generics: []
      }
    ]
  }
]
```

So in here, the argument with name `""` is a "pure" generic.

Comment (camelid) [reply src/librustdoc/formats/item_type.rs:51]: But value parameters like `a: T` don't have item types (at least from looking at the `ItemType` enum definition), so how are they represented?

Comment (GuillaumeGomez) [reply src/librustdoc/formats/item_type.rs:51]: Just with an empty name. If they are still present, might be interesting to remove them since they're useless for the search index.

File: src/librustdoc/html/render/cache.rs (hunk 5 above)
Comment (camelid) [reply src/librustdoc/html/render/cache.rs:302]: I just discussed this with Guillaume, and I think I can explain it now.

For purposes of search-indexing, `fn foo<T: Display>(x: T)` is treated as the type `fn(x: Display)`. This makes a search like `trait:Display ->` show `foo`. (It turns out you can filter by item type in rustdoc searches, which is part of what `ItemType` is for.)

However, `fn foo<T: Display + Clone>(x: T)`'s argument *cannot* be reduced into an `ItemType::Trait` because there are two traits involved, the `Display` trait and the `Clone` trait. So instead it's represented as a `ItemType::Generic`, which means there was more than one bound involved.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:302]: Exactly. :)

**Comment by jyn514** [comment]
Oh, sorry, I just saw this ping. Do you still need me to review?

**Comment by GuillaumeGomez** [comment]
If you have time, it'd very appreciated, otherwise I'll try to help @camelid finish it.

**Comment by jyn514** [comment]
It sounds like Camelid understands what's going on better than me, I'll leave them assigned.

File: src/librustdoc/html/render/cache.rs [hunk 9]
```diff
@@ -357,35 +374,35 @@ crate fn get_all_types<'tcx>(
...
     for arg in decl.inputs.values.iter() {
         if arg.type_.is_self_type() {
             continue;
         }
-        let mut args = FxHashSet::default();
+        args.clear();
         get_real_types(generics, &arg.type_, tcx, 0, &mut args);
         if !args.is_empty() {
-            all_types.extend(args);
+            all_types.extend(args.drain(..));
```
Comment (camelid) [inline src/librustdoc/html/render/cache.rs:387]: Why did you move `let mut args` outside of the loop and add `.clear()` and `.drain(..)`? Is there a behavioral difference between the old and new code?

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:387]: There are two things happening here:

 * Before you couldn't have the same type multiple times (because of the hashset), which is problematic when you're looking for `u32, u32` for example (function taking two `u32` for example).
 * As for the `clear` and `drain`, it's simply to be sure that the `args` vector only contains what we're looking for and improve memory usage.

Comment (camelid) [reply src/librustdoc/html/render/cache.rs:387]: Could you remove the second change and instead have the `let mut args = vec![];` inside the loop? I know it might not be as performant, but having the state persist across iterations makes this change harder to review. I'd rather not mix performance and behavior changes too.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:387]: Is it ok if I split it into two commits?

Comment (camelid) [reply src/librustdoc/html/render/cache.rs:387]: I'd rather it be a separate PR; I see some potential edge cases with it, and the slight performance improvement doesn't seem worth it to me (at least for now) for the additional code complexity.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:387]: I'll add a FIXME then.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs]: Done!

File: src/librustdoc/html/render/cache.rs [hunk 10]
```diff
@@ -357,35 +374,36 @@ crate fn get_all_types<'tcx>(
...
             continue;
         }
-        let mut args = FxHashSet::default();
+        // FIXME: performance wise, it'd be much better to move `args` declaration outside of the
+        // loop and replace this line with `args.clear()`.
+        let mut args = Vec::new();
         get_real_types(generics, &arg.type_, tcx, 0, &mut args);
         if !args.is_empty() {
-            all_types.extend(args);
+            all_types.extend(args.drain(..));
```
Comment (camelid) [inline src/librustdoc/html/render/cache.rs:388]: I think this `drain` can be removed now too, right?

```suggestion
            all_types.extend(args);
```

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:388]: Yes... T_T

Saying my good bye to perf improvements. :'(

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
ping @camelid

File: src/librustdoc/html/render/cache.rs [hunk 11]
```diff
@@ -194,32 +194,24 @@ crate fn get_index_search_type<'tcx>(
...
-        .collect::<Vec<_>>();
+    inputs.retain(|a| a.ty.name.is_some());
+    output.retain(|a| a.ty.name.is_some());
     let output = if output.is_empty() { None } else { Some(output) };
 
     Some(IndexItemFunctionType { inputs, output })
 }
 
-fn get_index_type(clean_type: &clean::Type) -> RenderType {
+fn get_index_type(clean_type: &clean::Type, generics: Vec<TypeWithKind>) -> RenderType {
```
Comment (camelid) [inline src/librustdoc/html/render/cache.rs:211]: Why did you remove `get_generics` and change this function to take the `generics` as an argument? I don't see any changes that would make that necessary.

File: src/librustdoc/html/render/cache.rs [hunk 12]
```diff
@@ -357,35 +374,38 @@ crate fn get_all_types<'tcx>(
...
+                    ret_types.push(TypeWithKind::from((get_index_type(return_type, vec![]), kind)));
                 }
             }
-            ret.into_iter().collect()
         }
-        _ => Vec::new(),
+        _ => {}
     };
-    (all_types.into_iter().collect(), ret_types)
+    (all_types, ret_types)
```
Comment (camelid) [inline src/librustdoc/html/render/cache.rs:410]: Is this the reason you changed the `FxHashSet`s to `Vec`s?

File: src/librustdoc/html/static/js/search.js [hunk 13]
```diff
@@ -345,6 +345,7 @@ window.initSearch = function(rawSearchIndex) {
         // Check for type name and type generics (if any).
         function checkType(obj, val, literalSearch) {
             var lev_distance = MAX_LEV_DISTANCE + 1;
+            var tmp_lev = MAX_LEV_DISTANCE + 1;
```
Comment (camelid) [inline src/librustdoc/html/static/js/search.js:348]: Why this `tmp_lev` change? Is it really necessary for the generics fix? IIUC, the `[NAME]` additions below are necessary, but I don't see why the `tmp_lev` addition is.

File: src/librustdoc/html/static/js/search.js [hunk 14]
```diff
@@ -404,17 +405,17 @@ window.initSearch = function(rawSearchIndex) {
                 // a levenshtein distance value that isn't *this* good so it goes
                 // into the search results but not too high.
                 lev_distance = Math.ceil((checkGenerics(obj, val) + lev_distance) / 2);
-            } else if (obj.length > GENERICS_DATA && obj[GENERICS_DATA].length > 0) {
+            }
+            if (obj.length > GENERICS_DATA && obj[GENERICS_DATA].length > 0) {
```
Comment (camelid) [inline src/librustdoc/html/static/js/search.js:409]: Same here. Why is this necessary? I'm not comfortable reviewing changes to the JS search code other than adding `[NAME]` as necessary.

**Comment by camelid** [comment]
Please also squash your commits together. IIUC, the test update is logically part of the other changes because the testsuite wouldn't pass without the test being updated.

File: src/librustdoc/html/render/cache.rs (hunk 11 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:211]: Because otherwise you need to set it into the returned type. With this, it cannot be forgotten.

File: src/librustdoc/html/render/cache.rs (hunk 12 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:410]: If by "this" you mean the fact that a same type can be present multiple times into generics, then yes. ;)

File: src/librustdoc/html/static/js/search.js (hunk 13 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:348]: True, I'll revert this one.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:348]: Ah no, nevermind! So originally it was declared just below ([here](https://github.com/rust-lang/rust/pull/88268/files#diff-766c50b21af39c8c0abd3fdba5f3c49a73c38b7b66ded4c7939e83b7244dfd3dL385)), but since it's now being used [here](https://github.com/rust-lang/rust/pull/88268/files#diff-766c50b21af39c8c0abd3fdba5f3c49a73c38b7b66ded4c7939e83b7244dfd3dR413) as well, I needed to have the declaration at the top of the function.

File: src/librustdoc/html/static/js/search.js (hunk 14 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:409]: Because before, if we found that the item had a "close enough" name to what we're looking for, it then ran checks on generics as well, otherwise it was going to look into the generics.

Now, in both cases we go look into generics and then we pick the best match out of the two.

And yes, sorry for the JS changes... If that can make you feel better, you can always confirm it's working as expected by running the JS search tests.

**Comment by GuillaumeGomez** [comment]
Squashed the commits and answered your comments!

File: src/librustdoc/html/static/js/search.js (hunk 14 above)
Comment (camelid) [reply src/librustdoc/html/static/js/search.js:409]: The way I interpret your description is that it's not necessary to bundle the JS change with the `html/formats/cache.rs` change; it's a separate bugfix. If so, can you please move this change to a separate PR and assign someone who's more knowledgeable about `search.js` on it? Or ask them to review the JS part of this PR?

> And yes, sorry for the JS changes... If that can make you feel better, you can always confirm it's working as expected by running the JS search tests.

Rustdoc doesn't currently have very exhaustive test coverage, so I'd rather have someone more knowledgeable look over the JS change.

File: src/librustdoc/html/static/js/search.js (hunk 13 above)
Comment (camelid) [reply src/librustdoc/html/static/js/search.js:348]: I don't really understand why the other changes were made. Like with https://github.com/rust-lang/rust/pull/88268#discussion_r711662034, can you please have someone who's more knowledgeable about rustdoc's JS review this part?

File: src/librustdoc/html/render/cache.rs (hunk 11 above)
Comment (camelid) [reply src/librustdoc/html/render/cache.rs:211]: I'm not sure what you mean. I'd like this change to be as small and simple as possible since I haven't interacted with this part of rustdoc before. I'd really appreciate it if you could revert all changes in this PR other than the part that inlines bounds better, because it's really hard for me to understand what the changes are doing when they are all mixed together.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:211]: All the changes are about that, including this one...

File: src/librustdoc/html/static/js/search.js (hunk 13 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:348]: I'll ask @Nemo157.

File: src/librustdoc/html/static/js/search.js [hunk 15]
```diff
@@ -404,17 +405,17 @@ window.initSearch = function(rawSearchIndex) {
...
                 // We can check if the type we're looking for is inside the generics!
                 var olength = obj[GENERICS_DATA].length;
                 for (x = 0; x < olength; ++x) {
-                    lev_distance = Math.min(levenshtein(obj[GENERICS_DATA][x], val.name),
-                                            lev_distance);
+                    tmp_lev = Math.min(levenshtein(obj[GENERICS_DATA][x][NAME], val.name), tmp_lev);
```
Comment (Nemo157) [inline src/librustdoc/html/static/js/search.js:413]: Wouldn't we need to recurse at this point to check whether the generics of the generic match the searched generics?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:413]: Absolutely, it's just that I wanted to keep the current (broken) behaviour for the moment. I plan to add it into the rework of the search engine. But if you prefer that I make this change now, I can as well.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:413]: Done!

**Review by notriddle — CHANGES_REQUESTED** [review]
Other than the naming thing, I really have only one thing to ask. The code changes, in substance, look reasonable to me.

Can you write some test cases for the JavaScript changes?

**Comment by GuillaumeGomez** [comment]
I added the test you suggested @notriddle (in `src/test/rustdoc-js/generics.*`) and improved the naming as suggested. :)

**Review by notriddle — APPROVED** [review]
